service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedResourceLimitChecks.java

```diff
@@ -237,34 +267,55 @@ long getMaximumNumberOfBytes(final TenantObject tenant) {
     }
 
     /**
-     * Gets the number of days for which the data usage is calculated and compared against the
-     * {@link #getMaximumNumberOfBytes(TenantObject)}.
+     * Gets the mode of the calculation. The supported modes are <em>days</em> and <em>monthly</em>.
      *
      * @param tenant The tenant configuration object.
-     * @return The number of days for which the data usage is calculated or {@link #DEFAULT_PERIOD_IN_DAYS} if not set.
+     * @return The mode of the calculation or {@code null} if not set.
```

Comment:
I remember in the past we decided to keep it away from the `TenantObject`. `TenantObject` offers methods to set and get the `resource-limits` and it doesn't worry what is inside the `resource-limits` configuration. IMHO the advantage of this approach is, that any custom implementations of the `ResourceLimitChecks` are now free to decide the configuration parameters based on their needs. May be a custom implementation doesn't need any of the configuration parameters that we have defined already.

Reply:
Hmmm, that makes sense to some degree. However, the Tenant API already defines most of the particular properties that we use for enforcing message limits. If we want to follow the approach you outlined above, then IMHO we would need to remove these properties from the Tenant API spec and just mention, that the Tenant API may return additional properties that can be used by the adapters for such purposes.

However, since we have implemented message limit enforcement as a standard functionality in the adapters, I do not see why we wouldn't/shouldn't explicitly define and explain these parameters. Or am I mistaken? Otherwise, where would be the right place to document the parameters expected by the adapters?

Reply:
> Hmmm, that makes sense to some degree. However, the Tenant API already defines most of the particular properties that we use for enforcing message limits. If we want to follow the approach you outlined above, then IMHO we would need to remove these properties from the Tenant API spec and just mention, that the Tenant API may return additional properties that can be used by the adapters for such purposes.

> 
> However, since we have implemented message limit enforcement as a standard functionality in the adapters, I do not see why we wouldn't/shouldn't explicitly define and explain these parameters. Or am I mistaken? Otherwise, where would be the right place to document the parameters expected by the adapters?

The limit enforcement in the adapters just  take _tenantObject_ as a parameter and doesn't worry about what is inside. What lies inside is taken care by the underlying `ResourceLimitChecks` implementation. I would personally prefer removing these properties from TenantAPI spec and delegate these to the `ResourceLimitChecks` implementation. On the other hand, I don't see any other better place other than the TenantAPI to document.

IMHO it is also acceptable to keep these parameters  in Tenant API provided the _resource-limits_ section in the API offers provision to set and return addition properties which can be used by any custom `ResourceLimitChecks` implementation.

Reply:
> IMHO it is also acceptable to keep these parameters in Tenant API provided the resource-limits section in the API offers provision to set and return addition properties which can be used by any custom ResourceLimitChecks implementation.

I think so too. As a consequence, we could then indeed move the methods to `TenantObject`, as these properties are now *officially* required/supported, right?

The Management API already supports adding arbitrary additional properties by means of the *ext* field. We would then simply add that field to the Tenant API (and `TenantObject`) as well.  WDYT?

Reply:
> I think so too. As a consequence, we could then indeed move the methods to `TenantObject`, as these properties are now _officially_ required/supported, right?
> 
> The Management API already supports adding arbitrary additional properties by means of the _ext_ field. We would then simply add that field to the Tenant API (and `TenantObject`) as well. WDYT?

It sounds good. I will create an issue for the support of `extensions` in the Tenant API (`TenantObject`) and suggest a separate PR for that.

To continue with this PR, I would suggest moving the above mentioned methods to the `TenantObject`.

Reply:
@sophokles73 I have now added those methods to the `TenantObject`. Also updated the  Device Management API and Device Registry implementation where the `data-volume` configurations were missing.